include/base/ghosting_functor.h

```diff
@@ -159,11 +160,29 @@ class GhostingFunctor : public ReferenceCountedObject<GhostingFunctor>
    */
   GhostingFunctor() {}
 
+  /**
+   * Copy Constructor
+   */
+  GhostingFunctor(const GhostingFunctor & other) : ReferenceCountedObject<GhostingFunctor>(other) {}
+
   /**
    * Virtual destructor; this is an abstract base class.
    */
   virtual ~GhostingFunctor() {}
 
+  /**
+   * A clone() is needed because GhostingFunctor can not be shared between
+   * different meshes. The operations in  GhostingFunctor are mesh dependent.
+   */
+  virtual std::unique_ptr<GhostingFunctor> clone () const
+  { libmesh_not_implemented(); }
+
+  /**
+   * It should be called after cloning a ghosting functor.
+   * Ghosting functor is mesh dependent
+   */
+  virtual void set_mesh(const MeshBase * /*mesh*/) { libmesh_not_implemented(); }
```

Comment:
Let's just move the mesh pointer that so many subclasses use into this class, and then we can make this non-virtual?

Reply:
Well, not non-virtual, but at least we can give it a good default implementation.

Reply:
I intended to put `libmesh_not_implemented` to force us to update downstream but at the same time, I do not need to update all of them at once.

I am not sure what a good default implementation looks like. We may just move up mesh here since most of the functors are mesh-dependent. 

We might have something like this in `ghosting_funcgor.`h:?

```
virtual void set_mesh(const MeshBase * mesh) { _mesh = mesh; }

const mesh * getMesh() const { return _mesh; }

private:

const MeshBase * _mesh;
```

Reply:
Exactly.  But make it get_mesh; don't mix snake-case for the setter and camel-case for the getter!

We can then upgrade the libMesh subclasses with mesh pointers, and then updating MOOSE subclasses can come after the libMesh PR is merge.